halo1/model.py

```diff
@@ -46,27 +51,126 @@ def read_halo1model(filepath):
+
+	attach_bones is a tuple of prefixes that this function should attempt to
+	connect.
+
+	Returns the armature object and a dict of index - bone pairs.
 	'''
+	view_layer = bpy.context.view_layer
+
 	scene_nodes = dict()
+
+	absolute_transforms = get_absolute_node_transforms_from_jms(jms.nodes)
+
+	armature = bpy.data.armatures.new('imported')
+	armature_obj = bpy.data.objects.new('imported', armature)
+
+	# We need to specifically link the object to this so we can use mode_set
+	# to set Blender to edit mode.
+
+	view_layer.active_layer_collection.collection.objects.link(armature_obj)
+	view_layer.objects.active = armature_obj
+
+	# We need to be in edit mode in order to be able to edit armatures at all.
+
+	edit_bones = armature.edit_bones
+	bpy.ops.object.mode_set(mode='EDIT')
+
+	# "directions" of where each node is pointing for the attachment progress.
+	# These really are just points that are offset 1000 units in the direction
+	# the node is pointing from the point of the node's in scene head.
+	# These are used so we can check if two bones should be attached.
+	directions = {}
+
 	for i, node in enumerate(jms.nodes):
-		scene_node = create_sphere(
-			name=NODE_NAME_PREFIX+node.name, size=node_size)
+		scene_node = edit_bones.new(name=NODE_NAME_PREFIX+node.name)
 
 		# Assign parent if index is valid.
 		scene_node.parent = scene_nodes.get(node.parent_index, None)
 
-		set_rotation_from_jms(scene_node, node)
+		# If a node has multiple children we cannot connect its tail end
+		# to the children.
+
+		pos = absolute_transforms[i]['translation']
+		rot = absolute_transforms[i]['rotation']
+
+		# Bones when created start at 0 0 0 for their tail and head.
+		# We extend the tail so it sticks out a bit and signifies the direction.
+
+		scene_node.tail = (scene_node.bbone_x, 0.0, 0.0)
 
-		set_translation_from_jms(scene_node, node, scale)
+		# We then rotate the bone by the absolute rotation that it should have.
 
+		scene_node.transform(rot.to_matrix(), scale=False)
+
+		# Then we add the absolute location to both the head and the tail.
+
+		scene_node.head += Vector(
+			(pos[0] * scale, pos[1] * scale, pos[2] * scale))
+		scene_node.tail += Vector(
+			(pos[0] * scale, pos[1] * scale, pos[2] * scale))
+
+		# Store the info we gathered this cycle.
 		scene_nodes[i] = scene_node
 
-	return scene_nodes
+		# Store a "direction" for this node
+
+		direction = Vector((1000.0, 0.0, 0.0))
+		direction.rotate(rot)
+		directions[scene_node.name] = direction
+
+	if len(attach_bones) > 0:
+		print('Checking if any bones with the prefixes %s can be connected.\n' %
+				(attach_bones,))
+
+	# Attach all the bones that we should attach if we can safely do so.
+	for bone in armature.edit_bones:
+		children = bone.children
+
+		if (len(children) == 1 # Can't connect to multiple children, so don't.
+		and bone.name[len(NODE_NAME_PREFIX): ].startswith(attach_bones)
+		and children[0].name[len(NODE_NAME_PREFIX): ].startswith(attach_bones)):
```

Comment:
I understand only attaching when there's one child, but what is the rationale behind these other two checks? It seems like you're checking if the name starts with the certain string, but I don't follow why that's relevant to establish a link.

Remember you can use helper methods to pull out complex conditionals like this and give them an easy-to-understand name.

Reply:
This is done because some types of nodes we might not want to automatically connect. Like frames. So we define the wished behavior based on prefix.

Reply:
That's gotta be documented in there somewhere. Possibly as a helper function.

Reply:
That's documented in the docstring at the top of the function